### ISSUE TITLE
GCM doesn't work after mapbox update

### ISSUE THREAD

**Issue by Torvin** [microg/android_packages_apps_GmsCore#814]
After update to the mapbox version my microG crashed a couple of times and I discovered that it lost all its settings. I had to re-register the backends and re-enable GCM. And now it says 0 apps are registered and the state is 'disconnected'. It doesn't even show any existing apps as 'not resgistered'. It's just a blank screen:

![Screenshot_20190604-230821_microG_Services_Core](https://user-images.githubusercontent.com/3890785/58874897-29b4e380-871e-11e9-9aa6-4dc412c3fec6.png)

I cannot make any app to re-register by removing its com.google.android.gms.appid.xml like I used to before. Re-installing the app doesn't help either. Here's what Whatsapp says after removing the appid and restarting it:

```
                            Process com.whatsapp created for activity com.whatsapp/.Main
                            PID: 8421   UID:    GIDs: 

               MultiDex  I  VM with version 2.1.0 has multidex support
                         I  Installing application
                         I  VM has multidex support, MultiDex support library is disabled.
SigquitBasedANRDetector  D  Registered
         ActivityThread  D  Loading provider com.whatsapp.provider.contact: com.whatsapp.contact.ContactProvider
               zygote64  W  Class d.f.w.g failed lock verification and will run slower.
                         W  Common causes for lock verification issues are non-optimized dex code
                         W  and incorrect proguard optimizations.
            FirebaseApp  D  com.google.firebase.auth.FirebaseAuth is not linked. Skipping initialization.
                         D  com.google.firebase.crash.FirebaseCrash is not linked. Skipping initialization.
   FirebaseInitProvider  I  FirebaseApp initialization successful
                     FA  I  Collection disabled with firebase_analytics_collection_deactivated=1
                         I  App measurement is starting up, version: 11910
                         I  To enable debug logging run: adb shell setprop log.tag.FA VERBOSE
                         I  To enable faster debug mode event logging run:
                         I    adb shell setprop debug.firebase.analytics.app com.whatsapp
               zygote64  W  Class d.f.b.c failed lock verification and will run slower.
                         W  Class d.f.Z.Qa failed lock verification and will run slower.
  NetworkSecurityConfig  D  Using Network Security Config from resource network_security_config debugBuild: false
                  Event  D  No subscribers registered for event class d.f.C.c
                         D  No subscribers registered for event class e.a.a.h
                         D  No subscribers registered for event class d.f.C.c
                         D  No subscribers registered for event class e.a.a.h
  AppCompatViewInflater  I  app:theme is now deprecated. Please move to using android:theme instead.
                    NFC  V  this device does not have NFC support
         OpenGLRenderer  D  HWUI GL Pipeline
                 Adreno  I  QUALCOMM build                   : dceea4a, Iedf7298c8c
                         I  Build Date                       : 02/19/18
                         I  OpenGL ES Shader Compiler Version: EV031.22.00.01
                         I  Local Branch                     :
                         I  Remote Branch                    :
                         I  Remote Branch                    :
                         I  Reconstruct Branch               :
                         I  PFP: 0x005ff087, ME: 0x005ff063
               zygote64  I  android::hardware::configstore::V1_0::ISurfaceFlingerConfigs::hasWideColorDisplay retrieved: 0
         OpenGLRenderer  I  Initialized EGL, version 1.4
                         D  Swap behavior 2
               zygote64  I  Method exceeds compiler instruction limit: 29726 in int com.whatsapp.emoji.EmojiDescriptor.getDescriptor(com.whatsapp.emoji.EmojiDescriptor$b)
                         W  Long monitor contention with owner pool-1-thread-1 (8459) at long com.android.org.conscrypt.NativeCrypto.RSA_generate_key_ex(int, byte[])(NativeCrypto.java:-2) waiters=0 in java.security.KeyPair d.e.b.a.z.f(java.lang.String) for 525ms
           RecyclerView  E  No adapter attached; skipping layout
               zygote64  I  Do partial code cache collection, code=2912B, data=30KB
                         I  After code cache collection, code=2912B, data=30KB
                         I  Increasing code cache capacity to 128KB
     FirebaseInstanceId  E  Token retrieval failed: SERVICE_NOT_AVAILABLE
 SpannableStringBuilder  E  SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
                         E  SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
                         E  SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
                         E  SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
```

**Comment by Nanolx** [comment]
GCM is not related to Mapbox (which is for Maps API), do you have any adblocker blocking `mtalk.google.com`?

Though it looks like WhatsApp is now using FCM only, because it doesn't show anything GCM related in your log.

As you may have heard GCM is beeing superseded by FCM and while it still works, support for GCM officially ended in May. So more and more apps are dropping GCM support in favor of FCM. microG has partial FCM support, not yet complete - that might be the cause.

**Comment by Torvin** [comment]
I realise that mapbox itself has nothing to do with it, I'm just referring to version `0.2.7.17455-mapbox` from FDroid, after installing which I started to have this issue. I'm pretty sure that version includes other maps-unrelated changes, which caused the bug. Sorry for the confusion.

`mtalk.google.com` is not blocked - I can ping it from the device and it's not listed in `/etc/hosts`

Here's another log from Twidere:

```
                            Process org.mariotaku.twidere created for activity org.mariotaku.twidere/.activity.MainActivity
                            PID: 9667   UID:    GIDs: 

               MultiDex  I  VM with version 2.1.0 has multidex support
                         I  Installing application
                         I  VM has multidex support, MultiDex support library is disabled.
            FirebaseApp  D  com.google.firebase.auth.FirebaseAuth is not linked. Skipping initialization.
                         D  com.google.firebase.crash.FirebaseCrash is not linked. Skipping initialization.
                     FA  I  App measurement is starting up, version: 11400
                         I  To enable debug logging run: adb shell setprop log.tag.FA VERBOSE
                         I  To enable faster debug mode event logging run:
                         I    adb shell setprop debug.firebase.analytics.app org.mariotaku.twidere
   FirebaseInitProvider  I  FirebaseApp initialization successful
  NetworkSecurityConfig  D  No Network Security Config specified, using platform default
         WebViewFactory  I  Loading com.android.webview version 72.0.3626.121 (code 362612150)
       cr_LibraryLoader  I  Time to load native libraries: 3 ms (timestamps 5584-5587)
               chromium  I  [INFO:library_loader_hooks.cc(42)] Chromium logging enabled: level = 0, default verbosity = 0
       cr_LibraryLoader  I  Expected native library version number "72.0.3626.121", actual native library version number "72.0.3626.121"
         cr_ChildProcLH  W  Create a new ChildConnectionAllocator with package name = com.android.webview, sandboxed = true
      cr_BrowserStartup  I  Initializing chromium process, singleProcess=false
        riotaku.twidere  W  type=1400 audit(0.0:299): avc: denied { read } for name="vmstat" dev="proc" ino=4026532131 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:proc:s0 tclass=file permissive=0
    cr_CrashFileManager  W  /data/user/0/org.mariotaku.twidere/cache/WebView/Crash Reports does not exist or is not a directory
         DynamiteModule  W  Failed to retrieve remote module version.
               zygote64  I  The ClassLoaderContext is a special shared library.
  GmsDynamiteLoaderImpl  D  unimplemented Method: getModuleVersion for com.google.android.gms.ads.dynamite
            ContextImpl  W  Failed to ensure /data/user/0/com.google.android.gms/shared_prefs: mkdir failed: EACCES (Permission denied)
                         W  Failed to ensure /data/user/0/com.google.android.gms/shared_prefs: mkdir failed: EACCES (Permission denied)
      AppCompatDelegate  I  The Activity's LayoutInflater already has a Factory installed so we can not install AppCompat's
        TwilightManager  I  Could not get last known location. This is probably because the app does not have any location permissions. Falling back to hardcoded sunrise/sunset values.
                     FA  I  Tag Manager is not found and thus will not be used
      AppCompatDelegate  I  The Activity's LayoutInflater already has a Factory installed so we can not install AppCompat's
                    NFC  V  this device does not have NFC support
        TwilightManager  I  Could not get last known location. This is probably because the app does not have any location permissions. Falling back to hardcoded sunrise/sunset values.
         OpenGLRenderer  D  HWUI GL Pipeline
                 Adreno  I  QUALCOMM build                   : dceea4a, Iedf7298c8c
                         I  Build Date                       : 02/19/18
                         I  OpenGL ES Shader Compiler Version: EV031.22.00.01
                         I  Local Branch                     :
                         I  Remote Branch                    :
                         I  Remote Branch                    :
                         I  Reconstruct Branch               :
                         I  PFP: 0x005ff087, ME: 0x005ff063
               zygote64  I  android::hardware::configstore::V1_0::ISurfaceFlingerConfigs::hasWideColorDisplay retrieved: 0
         OpenGLRenderer  I  Initialized EGL, version 1.4
                         D  Swap behavior 2
          Choreographer  I  Skipped 43 frames!  The application may be doing too much work on its main thread.
```

Both `com.google.android.gms.appid.xml` and `com.google.android.gms.measurement.prefs.xml` were sucessfully regenerated after this. Still the app list in microG is empty. Also those `EACCES` errors are concerning.

Let me clarify. Before upgrading to `0.2.7.17455-mapbox` there were around 30 apps listed in microG. Some of them were listed as 'not registered' (like Steam for example, #737). After installing it microG lost its settings and now there are zero. The screen is **empty**. I guess the questions that I have are:

- Why did microG lose its settings? Is it a bug, or was that suppose to happen?
- Why is the 'Apps using GCM' list is empty? Previously it included the apps that failed to register. Did the algorithm that builds this list change or is this a bug?
- Why does the status is shown as 'Disconnected'? Which logs can I look at to find the root cause?

**Comment by Eerovil** [comment]
I think I had the same problem (GCM Disconnected, same version). I did this to fix it:

1. Remove google account from settings/accounts
2. Uninstall all GCM using applications
3. Disable/Enable GCM for good measure
4. Reboot
5. Disable/Enable GCM for good measure again
6. Install an app that uses GCM.

**Comment by Torvin** [comment]
@Eerovil thank you! Unfortunately that doesn't work for me - I don't have a google account and re-installing apps doesn't help at all...

**Comment by Eerovil** [comment]
@Torvin Right, I guess it's something else then. You probably tried clearing the app data for GmsCore + Rebooting?

**Comment by Torvin** [comment]
@Eerovil Thank you so much! I tried that again and it worked this time!

**Comment by Eerovil** [comment]
Great! It's possible that the fix described in #822 would have also helped.

**Comment by Torvin** [comment]
@Eerovil nope, I tried that without any luck. Thank you again!
